### ISSUE TITLE
yolov8 tflite infer result error

### ISSUE THREAD

**Issue by vonchenplus** [ultralytics/ultralytics#4995]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

I am training a detection model and attempting to export it as a TFLite model for inference. However, I have noticed significant differences in its output compared to the PT or ONNX formats, and it appears to be incorrect.

Export command:  _**yolo detect export model=best.pt format=tflite**_

Infer comnand: **_yolo detect predict model=best.onnx source=test.jpeg_**
PT Or Onnx output:
![image](https://github.com/ultralytics/ultralytics/assets/3349963/728459bb-ebf5-41d9-a292-ad7d353bd239)

Infer comnand: **_yolo detect predict model=best.tflite source=test.jpeg_**
Tflite Or saved_model output:
![image](https://github.com/ultralytics/ultralytics/assets/3349963/984df978-3168-4236-b980-1c3e7f695484)

### Environment

Ultralytics YOLOv8.0.170 🚀 Python-3.8.16 torch-1.13.1+cu117 CUDA:0 (NVIDIA A100-SXM4-40GB, 40415MiB)
Setup complete ✅ (56 CPUs, 251.8 GB RAM, 265.7/949.3 GB disk)

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@vonchenplus 👋 hi, thanks for letting us know about this possible problem with YOLO 🚀.

TFLite export and inference works correctly for me, I'm unable to reproduce any problems using the inference commands provided at the end of export:

<img width="1120" alt="Screenshot 2023-09-20 at 16 24 59" src="https://github.com/ultralytics/ultralytics/assets/26833433/63d2dc9d-770e-4cb1-aacf-03da12287170">

 We've created a few short guidelines below to help users provide what we need in order to start investigating a possible problem.

## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by vonchenplus** [comment]
Hello @glenn-jocher 

Thanks for you reponse.  

My operating steps are consistent with yours. I have uploaded my PyTorch model and test images. If you have the time, please test them.

https://drive.google.com/file/d/1VIxQHiWDbz3Y7eWq_x-ChyCwDq31HqXL/view?usp=sharing

**Comment by glenn-jocher** [comment]
Hello @vonchenplus,

Thank you for your patience and for sharing the resources needed to reproduce and investigate this issue. I would suggest a few things to help understand this situation better:

1. Please ensure that your training environment is set up correctly and matches the original YOLOv8 environment. This includes using the correct versions of PyTorch, other libraries and dependencies.
2. During the training phase, keep an eye on the model's performance and losses. If your model is not training properly, it might also result in less accurate exported tflite models.

It's interesting that you're finding the exported TFLite model normal with the pre-trained model. This could imply that the problem might be localized either during your training phase or in the specific configuration or data you're using.

Continue to monitor your metrics and diagnostics during training, and feel free to provide any additional information that might help us better understand your situation. 
Again, thank you for using YOLOv8 and for bringing this to our attention.

**Comment by RizwanMunawar** [comment]
@vonchenplus! The results have appeared fine to me (for PyTorch and Onnx Model). I have evaluated the model using the described procedure.

### PyTorch Inference

```
yolo task=detect mode=predict source=1695114798-1292956132.jpeg model=best.pt save=True
```

![1695114798-1292956132](https://github.com/ultralytics/ultralytics/assets/62513924/41f2c5e4-ad6f-42f1-b380-d911148abc00)

### Onnx Conversion and Inference
```
yolo export model=best.pt format=onnx
yolo task=detect mode=predict source=1695114798-1292956132.jpeg model=best.onnx save=True
```
![1695114798-1292956132](https://github.com/ultralytics/ultralytics/assets/62513924/a97816cb-a7ab-4579-a3ae-8892a92d224d)

### TensorFlow Lite
It appears that the model's training for various classes may not be optimal. Could you please specify how many classes are present in your dataset?

**Comment by vonchenplus** [comment]
Hello @RizwanMunawar 

Thanks for your response. 

I confirmed in the problem description that the pytorch and onnx models are working fine, but tflite is getting just plain wrong results. 

My training set contains a total of 16 classifications.

**Comment by RizwanMunawar** [comment]
@vonchenplus, I've accomplished the tflite export of `yolov8m.pt`, and its inference results match those of the PyTorch model. This suggests that any issues likely stem from your model training process. Additionally, I'd like to highlight that the `onnx2tf` script is operating as anticipated.

**Comment by vonchenplus** [comment]
Hello @glenn-jocher And @RizwanMunawar ，

Are you suggesting that there might be an issue with my training process? Which metrics could potentially help address this problem? My understanding is that if the PyTorch model is fine, the converted TFLite model should also function correctly.

**Comment by glenn-jocher** [comment]
Hello @vonchenplus,

Yes, an issue in the training process could lead to a discrepancy between your TFLite model and PyTorch model. Keep an eye for signs of overfitting in your model during the training process such as if your model is performing extremely well on the training set but not so well on the validation set. This could explain the high confidence scores you're seeing.

Verify the quality and diversity of your dataset. Make sure its well balanced among the 16 classes. Also, ensure your training configuration is appropriate for your specific problem and dataset. 

It's also useful to examine overall training and validation losses, as well as class-specific Precision, Recall and mAP metrics. These could give you insights about the generalization capability of your model.

Remember that conversion from PyTorch to TFLite involves optimizations that could lead to small discrepancies, but should not cause the complete absence of correct predictions.

**Comment by vonchenplus** [comment]
Hello @glenn-jocher,

After testing, it is most likely caused by overfitting. Thank you very much for your responses.

**Comment by RizwanMunawar** [comment]
Hello @vonchenplus!

We will conduct additional experiments with the `tflite export` module. I'll keep you updated with any new developments related to this issue. For the time being, I'm going to close this issue.

Thanks
Ultralytics Team!
